Document amoc coordinator (#102)
esl/amoc: doc/amoc_coordinator.md

# amoc_coordinator

## Module

`amoc_coordinator` 

## Description

This module allows to synchronize the users, and act on groups of them.

> The coordinator reacts to new users showing up in a system, according to `Coordination Plan`. `Coordination Plan` consists of `Coordination Items`, each of them is defined as one of the following: `{NumberOfUsers, CoordinationActions}`. When `Number Of Users` is set to `all`, then only `Coordination Actions` with the arities `/1, /2` are handled. When `Number Of Users` is set to a positive integer, then when then all `Coordination Actions` with the arities `/1, /2` and `/3` are handled. When new batch size `NumberOfUsers` of users calls `add` to the coordinator, then `Coordination Actions` following the `NumberOfUsers` in the `Coordination Plan` are triggered. When there is more then one `Coordination Items` matching the `NumberOfUsers` all are triggered, Each will be passed respective number of users, eg if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on 6th users calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.

Reviewer: Also this paragraph is a bit hard to read, more things were uncovered for me when reading the example. Do you think this paragraph could be reworked?

 - when `{A, A}` is not in the `distinct pairs`;
 - all pairs are distinct;
 - Eg. for `[a]` `distinct pairs` collection is `[{a, undefined}]`;
 - Eg. for `[a, b]` `distinct pairs` collection is `[{a, b}]`;
 - Eg. for `[a, b, c]` `distinct pairs` collection is `[{a, b}, {a, c}, {b, c}]`.

## Exports

### `start(CoordinatorName, CoordinationPlan) -> ok | error`
### `start(CoordinatorName, CoordinationPlan, Timeout) -> ok | error`

#### Types
```erlang
CoordinatorName :: atom()
CoordinationPlan  :: [ CoordinationItem ]
CoordinationItem :: {NumberOfUsers :: pos_integer() | all, CoordinationActions}
CoordinationActions :: [ CoordinationActions ] | CoordinationAction
CoordinationAction :: 
    fun(Event) -> any() |
    fun(Event, ListOfUsersData) -> any() |
    fun(Event, MaybeUser, MaybeUser) -> any()
Event :: coordinate | timeout | stop | reset
ListOfUsersData :: [ User ]
MaybeUser :: undefined | User
User :: {UsersPid :: pid(), Data :: any()}
Timeout :: pos_integer() | infinity
```

This function starts a coordinator. Usually is called in `init/0` of a amoc scenario.

### `add(CoordinatorName, Data) -> ok`
### `add(CoordinatorName, Pid, Data) -> ok`

> ### Types

Author: Yes

```erlang
CoordinatorName :: atom()
UsersPid :: pid()
Data :: any()
```

This function adds current process data. Usually is called in `start/2` of a amoc scenario.

## Example

This scenario will demonstrate how do the `users` interact with `amoc_coordinator`:

```erlang
-module(example).

-export([init/0]).
-export([start/2]).

init() ->
    Plan = [
        {2, fun(Event) ->
                io:fwrite("Two new users showed up: Event = ~p\n", [Event])
            end},
        {2, fun(Event, ListOfUsers) ->
                io:fwrite("Two new users showed up: Event = ~p; ListOfUsers = ~p\n", [Event, ListOfUsers]),
                [ Pid ! {hello, Data} || {Pid, Data} <- ListOfUsers]
            end},
        {2, fun(Event, User1, User2) ->
                io:fwrite("Two new users showed up: Event = ~p; User1 = ~p; User2 = ~p\n", [Event, User1, User2])
            end},
        
        {3, fun(_Event) ->
                io:fwrite("Three new users showed up\n", [])
            end},
        {all, fun(Event) ->

>                 io:fwrite("All users have logged in Event = ~p\n", [Event])

Author: It means, all users called `amoc_coordinator:add/2` or `amoc_coordinator:add/3`.
I shall replace `logged` with `called amoc_coordinator:add`.

            end}
        
    ],
    Settings = [setting1, {setting2, something}],
    amoc_coordinator:start(?MODULE, Plan),
    {ok, Settings}.

start(Id, _Settings) ->
    io:fwrite("User = ~p\n", [Id]),
    amoc_coordinator:add(?MODULE, Id),
    receive
        Msg -> 
            io:fwrite("{Msg = ~p, Id = ~p\n", [Msg, Id])
    end,
    ok.
```

To run it:

```bash
$ make rel
$ _build/default/rel/amoc/bin/amoc console

1> amoc_local:do(example, 1, 5).
```

Filtered, formated and explained output:

```erlang
User = 1  % First user is started

ok  % result of calling amoc_local:do/3

User = 2 % First user is started

% We have 2 users added to amoc_coordinator so all of actions {2, _} are triggered:
Two new users showed up: Event = coordinate; User1 = {<0.1142.0>,2}; User2 = {<0.1140.0>,1}
% This action triggers sending {hello,Id} to the users 1 and 2
Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1142.0>,2},{<0.1140.0>,1}]

> Two new users showed up: Event = coordinate

Author: The order of executions is always reversed with reference to order on the `CoordinationPlan`, I shall add a piece of information about that.

> Two new users showed up: Event = coordinate

Author: I do not know why is that.

 - Eg. for `[a]` `distinct pairs` collection is `[{a, undefined}]`;
 - Eg. for `[a, b]` `distinct pairs` collection is `[{a, b}]`;
 - Eg. for `[a, b, c]` `distinct pairs` collection is `[{a, b}, {a, c}, {b, c}]`.

## Exports

#### `start(CoordinatorName, CoordinationPlan) -> ok | error`
#### `start(CoordinatorName, CoordinationPlan, Timeout) -> ok | error`

##### Types
```erlang
CoordinatorName :: atom()
CoordinationPlan  :: [ CoordinationItem ]
CoordinationItem :: {NumberOfUsers :: pos_integer() | all, CoordinationActions}
CoordinationActions :: [ CoordinationActions ] | CoordinationAction
CoordinationAction :: 
    fun(Event) -> any() |
    fun(Event, ListOfUsersData) -> any() |
    fun(Event, MaybeUser, MaybeUser) -> any()
Event :: coordinate | timeout | stop | reset
ListOfUsersData :: [ User ]
MaybeUser :: undefined | User
User :: {UsersPid :: pid(), Data :: any()}
Timeout :: pos_integer() | infinity
```

This function starts a coordinator. Usually is called in `init/0` of a amoc scenario.

#### `add(CoordinatorName, Data) -> ok`
#### `add(CoordinatorName, Pid, Data) -> ok`

##### Types
```erlang
CoordinatorName :: atom()
UsersPid :: pid()
Data :: any()
```

> This function adds current process data. Usually is called in `start/2` of a amoc scenario.

Reviewer: I think it's important to say that this function basically increases number of already started users. Without this the coordination actions may not be triggered, especially the one for all users. If for some reason one session doesn't call the function and we wanted to have 10 users in the test, then the rule for `all` will not be triggered. Am I right?

Author: No

Author: If we have 10 users in a test but only 5 of them calls the function `add`. and the event `timeout` happens action for `all` will be triggered for the 5 users who registered to the amoc coordinator.

# amoc_coordinator

## Module

`amoc_coordinator` 

## Description

> This module allows to synchronize the users, and act on groups of them.

Reviewer 2: ```suggestion
This module allows to synchronize the users and act on groups of them.
```

The coordinator reacts to new users showing up in a system, according to the `Coordination Plan`. 
The `Coordination Plan` consists of `Coordination Items`, and each of them is defined as one of the following: `{NumberOfUsers, CoordinationActions}`.
 - When the `Number Of Users` is set to `all`, then only `Coordination Actions` with the arities `/1, /2` are handled. 
 The `Coordination Items` with `all` are triggered by the `timeout` event. 
 - When the `Number Of Users` is set to a positive integer, all `Coordination Actions` with arities `/1, /2` and `/3` are handled.

> The timeout timer is reset by calling function `add`.

Reviewer 3: ```suggestion
The timeout timer is reset by calling the `add` function.
```

A new batch size is set in the `NumberOfUsers`. Each user in the batch calls the `add` function registering to the coordinator and triggering the `Coordination Plan`.
If more then one of the `Coordination Items` matching the `NumberOfUsers` is triggered, each of them will be passed a respective number of users.
For example if the `Coordination Plan` is `[{2, Act1}, {3, Act2}]` then on the 6th user calling `add`, `Act1` will be called with 2 users passed and `Act2` wil be called with 3 users passed.

`Coordination Actions` may be one of the following:
 - `fun(Event) -> any()` - this type of action does not care about particular users, but only about the number of them;
 - `fun(Event, ListOfUsersData) -> any()` - this type of action gets `ListOfUsersData` which is a list of tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;
- `fun(Event, User1, User2) -> any()` - this type of action gets `distinct pairs` from the batch of users `User1` and `User2` which are tuples `{Pid, Data}` with `Pid`s passed by users calling `amoc_coordinator:add/2` or `amoc_coordinator:add/3`;

> `Coordination Actions` are executed in reversed order in relation to thair declaration in the `Coordination Plan`.

Reviewer 3: ```suggestion
`Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.
```

> `Coordination Actions` are executed in a reversed order in relation to their declaration in the `Coordination Plan`.

Reviewer 4: this is not true. the order of actions in one item is guaranteed (direct order, not reversed). the order of actions from different coordination items is not guaranteed at all (casted and executed in different processes). timeout/reset/stop events are not casted but ‘called’, so for the timeout/reset/stop evenst it’s guaranteed that all the coordination items with `all` are executed after all numbered coordination items are done.

`distinct pairs` - in the context, these are pairs from collection where:
 - when `{A, B}` is in the `distinct pairs` then `{B, A}` is not;
 - when `{A, A}` is not in the `distinct pairs`;

>  - all pairs are distinct;

Author: ??

>  - all pairs are distinct;

Reviewer 5: I just found it quite redundant, but I actually have no better suggestion 🤷‍♂

 - Eg. for `[a]` `distinct pairs` collection is `[{a, undefined}]`;
 - Eg. for `[a, b]` `distinct pairs` collection is `[{a, b}]`;

>  - Eg. for `[a, b, c]` `distinct pairs` collection is `[{a, b}, {a, c}, {b, c}]`.

Reviewer 5: > – Eg. for `xxx`, the `distinct pairs` collection is `yyy`

(comma and article missing)

## Exports

#### `start(CoordinatorName, CoordinationPlan) -> ok | error`
#### `start(CoordinatorName, CoordinationPlan, Timeout) -> ok | error`

##### Types
```erlang
CoordinatorName :: atom()
CoordinationPlan  :: [ CoordinationItem ]
CoordinationItem :: {NumberOfUsers :: pos_integer() | all, CoordinationActions}
CoordinationActions :: [ CoordinationActions ] | CoordinationAction
CoordinationAction :: 
    fun(Event) -> any() |
    fun(Event, ListOfUsersData) -> any() |

>     fun(Event, MaybeUser, MaybeUser) -> any()

Reviewer 5: I understand that the two users given to this function are ensured to be different, right?
Then maybe you can write
```erlang
    fun(Event, MaybeUser1, MaybeUser2) when MaybeUser1 =/= MaybeUser2 -> any()
```

Reviewer 2: @NelsonVides this is a type definition, so the users (the instances of data) might actually be different, but the type is the same nonetheless.

Reviewer 2: ```
$ erl -pa _build/default/lib/amoc/ebin/
Erlang/OTP 21 [erts-10.3.5.6] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1]

Enabled docsh 0.7.1 from: /Users/erszcz/work/erszcz/docsh/_build/default/lib/docsh
Call h(docsh) for interactive help.

Eshell V10.3.5.6  (abort with ^G)
1> l(amoc_coordinator).
{module,amoc_coordinator}
2> h(amoc_coordinator).
Module description is not available.
```

:'(

```
ok
3> t(amoc_coordinator).
-type coordination_action() :: fun((coordination_event(),
                                    [coordination_data()]) -> any()) |
                               fun((coordination_event(),
                                    maybe_coordination_data(),
                                    maybe_coordination_data()) -> any()) |
                               fun((coordination_event()) -> any()).
-type coordination_actions() :: [coordination_action()] | coordination_action().
-type coordination_data() :: {pid(), Data :: any()}.
-type coordination_event() :: coordinate | timeout | stop | reset.
-type coordination_item() :: {NoOfUsers :: pos_integer() | all,
                              coordination_actions()}.
-type coordination_plan() :: [coordination_item()] | coordination_item().
-type coordination_timeout_in_sec() :: pos_integer() | infinity.
-type maybe_coordination_data() :: coordination_data() | undefined.
-type normalized_coordination_item() :: {NoOfUsers :: pos_integer() | all,
                                         [coordination_action()]}.
-type state() :: {worker, pid()} | {timeout, pid()}.

ok
4>
```

:+1:

            end}
        
    ],
    Settings = [setting1, {setting2, something}],
    amoc_coordinator:start(?MODULE, Plan),
    {ok, Settings}.

start(Id, _Settings) ->
    io:fwrite("User = ~p\n", [Id]),
    amoc_coordinator:add(?MODULE, Id),
    receive
        Msg -> 
            io:fwrite("{Msg = ~p, Id = ~p\n", [Msg, Id])
    end,
    ok.
```

To run it:

```bash
$ make rel
$ _build/default/rel/amoc/bin/amoc console

1> amoc_local:do(example, 1, 5).
```

Filtered, formated and explained output:

```erlang
User = 1  % First user is started

ok  % result of calling amoc_local:do/3

User = 2 % First user is started

% We have 2 users added to amoc_coordinator so all of actions {2, _} are triggered:
Two new users showed up: Event = coordinate; User1 = {<0.1142.0>,2}; User2 = {<0.1140.0>,1}
% This action triggers sending {hello,Id} to the users 1 and 2
Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1142.0>,2},{<0.1140.0>,1}]
Two new users showed up: Event = coordinate

% Users 1 and 2 received messages and print them:
{Msg = {hello,2}, Id = 2
{Msg = {hello,1}, Id = 1

User = 3

> % We have 3 users added to amoc_coordinator so all of actions {3, _} are triggered:

Reviewer 5: > so all of the `{3, _}` actions [...]

(ordering of the english genitive, and one article)
(same in the sentence a few lines below)

Three new users showed up

User = 4

> % We have 4 and 4 rem 2 == 0 therefore users added to amoc_coordinator so all of actions {2, _} are triggered:

Reviewer 5: Same as above

Two new users showed up: Event = coordinate; User1 = {<0.1144.0>,4}; User2 = {<0.1143.0>,3}
Two new users showed up: Event = coordinate; ListOfUsers = [{<0.1144.0>,4},{<0.1143.0>,3}]
Two new users showed up: Event = coordinate

{Msg = {hello,4}, Id = 4
{Msg = {hello,3}, Id = 3
User = 5

% You need to wait for a while, and ...

> % Timeout has been reached. Which triggers all of the Coordination Actions with remaining number of users remaining.

Reviewer 5: "remaining" twice? 😛 

>  % Timeout has been reached, wiich triggers all of the Coordination Actions with the remaining number of users.
